src/main/java/training/bank/Bank.java

```diff
@@ -0,0 +1,61 @@
+package training.bank;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.security.InvalidParameterException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+
+public class Bank {
+    private static final Logger LOGGER = LogManager.getLogger();
+    private HashMap<String, Account> accounts = new HashMap<>(); // Is this better than HashSet?
+    private HashSet<Transaction> transactions = new HashSet<>();
+
+
+    public void updateWithTransactionRecord(TransactionRecord record) {
+            this.addNewAccounts(record);
+            Transaction newTransaction = this.addTransaction(record);
+            newTransaction.getPayee().updateTransactions(newTransaction);
+            newTransaction.getPayer().updateTransactions(newTransaction);
+    }
+
+    private void addNewAccounts(TransactionRecord record) {
+        if (!accounts.containsKey(record.getPayee())) {
+            accounts.put(record.getPayee(), new Account(record.getPayee()));
+        }
+        if (!accounts.containsKey(record.getPayer())) {
+            accounts.put(record.getPayer(), new Account(record.getPayer()));
+        }
+    }
+
+    private Transaction addTransaction(TransactionRecord transactionRecord) { // Check implementation of HS.addAll for efficiency
+
+        // Is it bad practice to assume that debtor and creditor are already in accounts?
```

Comment:
In general, it's bad practice to make assumptions in your public methods (because they can be used wrongly by other code) but totally fine to make assumptions in your private methods (along with a suitable comment if necessary).